File: android/src/main/kotlin/io/adrop/adrop_ads/rewarded/FlutterAdropRewardedAd.kt
package io.adrop.adrop_ads.rewarded

import android.app.Activity
import android.content.Context
import android.graphics.Color
import io.adrop.adrop_ads.AdropAd
import io.adrop.adrop_ads.AdType
import io.adrop.adrop_ads.bridge.AdropChannel
import io.adrop.adrop_ads.bridge.AdropMethod
import io.adrop.ads.model.AdropErrorCode
import io.adrop.ads.rewardedAd.AdropRewardedAd
import io.adrop.ads.rewardedAd.AdropRewardedAdListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

class FlutterAdropRewardedAd(
    context: Context,
    unitId: String,
    requestId: String,
    messenger: BinaryMessenger
): AdropAd(), AdropRewardedAdListener {
    private var rewardedAd: AdropRewardedAd = AdropRewardedAd(context, unitId)
    private var adropEventListenerChannel: MethodChannel?
    private var originalStatusBarColor: Int = 0
    private var showActivity: Activity? = null

    init {
        rewardedAd.rewardedAdListener = this
        val channelName = AdropChannel.adropEventListenerChannelOf(AdType.Rewarded, requestId)
        adropEventListenerChannel = if (channelName != null) {
            MethodChannel(messenger, channelName)
        } else {
            null
        }
    }

    override fun load() {
        rewardedAd.load()
    }

    override fun show(activity: Activity) {
        showActivity = activity
        originalStatusBarColor = activity.window.statusBarColor
        rewardedAd.show(activity) { type, amount ->
            val args = mapOf("type" to type, "amount" to amount)
            adropEventListenerChannel?.invokeMethod(AdropMethod.HANDLE_EARN_REWARD, args)
        }
    }

    override fun destroy() {
        rewardedAd.destroy()
    }

    override fun onAdFailedToReceive(ad: AdropRewardedAd, errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAIL_TO_RECEIVE_AD, mapOf("errorCode" to errorCode.name))
    }

    override fun onAdReceived(ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, metadataOf(ad))
    }

    override fun onAdClicked(ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, metadataOf(ad))
    }

    override fun onAdImpression(ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, metadataOf(ad))
    }

    override fun onAdDidPresentFullScreen(ad: AdropRewardedAd) {
        if (ad.isBackfilled) {
            showActivity?.window?.statusBarColor = Color.BLACK
        }
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_PRESENT_FULL_SCREEN, metadataOf(ad))
    }

    override fun onAdDidDismissFullScreen(ad: AdropRewardedAd) {
        showActivity?.window?.statusBarColor = originalStatusBarColor
        showActivity = null
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_DISMISS_FULL_SCREEN, metadataOf(ad))
    }

    override fun onAdFailedToShowFullScreen(ad: AdropRewardedAd, errorCode: AdropErrorCode) {
        showActivity?.window?.statusBarColor = originalStatusBarColor
        showActivity = null
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAIL_TO_SHOW_FULL_SCREEN, mapOf("errorCode" to errorCode.name))
    }

    private fun metadataOf(ad: AdropRewardedAd): Map<String, Any?> {
        return mapOf(
            "creativeId" to ad.creativeId,
            "txId" to ad.txId,
            "campaignId" to ad.campaignId
        )
    }
}

File: ios/Classes/rewarded/FlutterAdropRewardedAd.swift
import Foundation
import Flutter
import AdropAds


class FlutterAdropRewardedAd: NSObject, AdropAd, AdropRewardedAdDelegate {

    private let messenger:FlutterBinaryMessenger
    private let adropEventListenerChannel: FlutterMethodChannel?
    private let requestId: String
    private let rewardedAd: AdropRewardedAd

    init(
        unitId: String,
        requestId: String,
        messenger: FlutterBinaryMessenger) {
            self.messenger = messenger
            self.requestId = requestId
            let methodChannelName = AdropChannel.adropEventListenerChannel(adType: AdType.rewarded, id: requestId)
            self.adropEventListenerChannel = methodChannelName != nil ? FlutterMethodChannel(name: methodChannelName!, binaryMessenger: messenger) : nil

            self.rewardedAd = AdropRewardedAd(unitId: unitId)
        }

    func load() {
        self.rewardedAd.delegate = self
        rewardedAd.load()
    }

    func show() {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene else {
            return
        }

        guard let viewController = windowScene.windows.first?.rootViewController else {
            return
        }

        rewardedAd.show(fromRootViewController: viewController) { [weak self] type, amount in
            guard let strongSelf = self else { return }
            strongSelf.adropEventListenerChannel?.invokeMethod(AdropMethod.HANDLE_EARN_REWARD,
                                                               arguments: ["type": type, "amount": amount])
        }
    }

    func onAdReceived(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, arguments: metadataOf(ad))
    }

    func onAdFailedToReceive(_ ad: AdropRewardedAd, _ errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAILED_TO_RECEIVE, arguments: ["errorCode":AdropErrorCodeToString(code: errorCode)])
    }


    func onAdImpression(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, arguments: metadataOf(ad))
    }

    func onAdClicked(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, arguments: metadataOf(ad))
    }

    func onAdWillPresentFullScreen(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.WILL_PRESENT_FULL_SCREEN, arguments: metadataOf(ad))
    }

    func onAdDidPresentFullScreen(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_PRESENT_FULL_SCREEN, arguments: metadataOf(ad))
    }

    func onAdWillDismissFullScreen(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.WILL_DISMISS_FULL_SCREEN, arguments: metadataOf(ad))
    }

    func onAdDidDismissFullScreen(_ ad: AdropRewardedAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_DISMISS_FULL_SCREEN, arguments: metadataOf(ad))
    }

    func onAdFailedToShowFullScreen(_ ad: AdropRewardedAd, _ errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAIL_TO_SHOW_FULL_SCREEN, arguments: ["errorCode": AdropErrorCodeToString(code: errorCode)])
    }

    private func metadataOf(_ ad: AdropRewardedAd) -> [String: Any] {
        return [
            "unitId": ad.unitId,
            "creativeId": ad.creativeId,
            "txId": ad.txId,
            "campaignId": ad.campaignId,
            "browserTarget": ad.browserTargetValue.rawValue
        ]
    }
}
